src/main/java/com/gradle/ProxyFactory.java

```diff
@@ -0,0 +1,108 @@
+package com.gradle;
+
+import org.gradle.api.Action;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.*;
+
+class ProxyFactory {
+
+    static <T> T createProxy(Object target, Class<T> targetInterface) {
+        return newProxyInstance(targetInterface, new ProxyingInvocationHandler(target));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T newProxyInstance(Class<T> targetInterface, InvocationHandler invocationHandler) {
+        return (T) Proxy.newProxyInstance(targetInterface.getClassLoader(), new Class[]{targetInterface}, invocationHandler);
+    }
+
+    private static class ProxyingInvocationHandler implements InvocationHandler {
+
+        private final Object target;
+
+        ProxyingInvocationHandler(Object target) {
+            this.target = target;
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) {
+            try {
+                Class<?>[] parameterTypes = method.getParameterTypes();
+                Method targetMethod = target.getClass().getMethod(method.getName(), convertTypes(parameterTypes, target.getClass().getClassLoader()));
+                Object result = targetMethod.invoke(target, toTargetArgs(args));
+                if (result == null || isJdkType(result.getClass())) {
+                    return result;
+                }
```

Comment:
❔ Did you decide caching the lookup wasn't worth it?

Reply:
I had a cache in some interim version but removed it to restore configuration cache compatibility. I don't think performance is a real concern here since these APIs are not called excessively.